### ISSUE TITLE
[Enhancement] Add "Copy tag" option in Matomo Tag Manager

### ISSUE THREAD

**Issue by SW-Vincent** [matomo-org/tag-manager#615]
Sometimes we need to build a really similar Tag / Trigger within Matomo TMS. When it happens, being able to just copy a tag would do a fine addition to Matomo Tag Manager features !

## Summary

Considering this feature, it would require to add a "copy" button on the side of each Tag as you can see in the following exemple : 

![image](https://user-images.githubusercontent.com/118832007/229561638-3eaef062-cbef-4836-b5fa-8f2789c93e33.png)

Which would lead to the following screen : 

![image](https://user-images.githubusercontent.com/118832007/229563414-b6a2f9df-1888-4265-8544-8e94916166d9.png)

It would create the same tag, named "Copy of MTM - myTagName" with either : 
 - the same triggers if "No" was selected on the previous screen 
 - the copy of the triggers if "Yes" was selected on the previous screen

Of course this goes with the possibility to copy triggers within the triggers section directly as well.

## Your Environment

* Matomo Version: Matomo Cloud

**Comment by snake14** [comment]
Hi @SW-Vincent . Thank you for taking the time to create this issue. It appears that it might be a duplicate of [issue 51](https://github.com/matomo-org/tag-manager/issues/51). However, that one might have gotten lost in the backlog. Also, since you provided some good UI suggestions, I'm going to keep this issue and mark it for prioritisation.

**Comment by Stan-vw** [comment]
Thanks for the suggestions @SW-Vincent. I'll use the word "Duplicate" instead of "Copy" since the feature sounds like a duplication feature, not a copying feature (which should have a paste action).*

Duplicating Tags makes sense to me, duplicating triggers also makes sense to me. However, I'm not seeing the clear logic in asking whether we should create duplicates of Triggers when a user is duplicating a Tag. 

Would this feature work well for you if it didn't include the pop-up with the question to duplicate the triggers?
I.e. the moment you click on Duplicate Tag/Trigger it immediately does just that, and not something else as well.

*Edit: I saw some other products call it "Make a copy", which I guess is clear enough as well.

**Comment by SW-Vincent** [comment]
It would be great already !

**Comment by MatomoForumNotifications** [comment]
This issue has been mentioned on **Matomo forums**. There might be relevant details there:

https://forum.matomo.org/t/duplicate-tags-or-triggers/40457/8

**Comment by mlisowsk** [comment]
Please implement also for on premise. Specifically I would like to copy tags and triggers from one container to another.
Alternatively let us duplicate an entire container including triggers and tags, for example as an option when creating a new container.

**Comment by heurteph-ei** [comment]
Hi @mlisowsk 
>Alternatively let us duplicate an entire container including triggers and tags, for example as an option when creating a new container.

This is easily doable:
Save the origin container via the export button, then use the import button in the target container...

**Comment by Chardonneaur** [comment]
I have a client who is asking the exact same feature, same for variables.

**Comment by Stan-vw** [comment]
@SW-Vincent and co, it'd be great to hear your thoughts on the following: is it fine to click "copy tag" and it will copy the tag within the existing container (within the same website), or would it be more powerful and save you a lot of time if it had the added ability to choose which website+container you want to copy this to?

It'll be extra work, but I'd like to know what the desired endgame is before we start, so we can create a proper UX rather than changes upon changes 😄

**Comment by SW-Vincent** [comment]
Hi @Stan-vw, what i had in mind was copying **within the same container** (although copying tags might have its uses).

**Comment by Neugart** [comment]
Hi @SW-Vincent 

> @SW-Vincent and co, it'd be great to hear your thoughts on the following: is it fine to click "copy tag" and it will copy the tag within the existing container (within the same website), or would it be more powerful and save you a lot of time if it had the added ability to choose which website+container you want to copy this to?
> 
> It'll be extra work, but I'd like to know what the desired endgame is before we start, so we can create a proper UX rather than changes upon changes 😄

I would also be happy to have the "copy" features since we have a lot of similar tags where the url varies only a bit. For us, it would be great to have both possibilities:
- copy within the same container
- copy in another container

Thanks!

**Comment by Stan-vw** [comment]
Keen to get your thoughts on something like this, that would potentially be reusable for container/tags/triggers/variables, but perhaps also for reports like funnels and goals.
![image](https://github.com/user-attachments/assets/5b249e7b-b1b2-42e5-b9ff-5bce2bb7e60c)

**Comment by Neugart** [comment]
Looks good!

**Comment by Chardonneaur** [comment]
I got a user asking for this feature.

**Comment by AltamashShaikh** [comment]
@Chardonneaur This issue is on our list of changes and its scheduled in our upcoming sprints

**Comment by Neugart** [comment]
Hello @Stan-vw ,

do you know when the feature will be available?
Thanks! :)

**Comment by snake14** [comment]
Hi @Neugart . I'm actually working on implementing this feature right now. It will likely be included in the next Matomo 5.2.0-x release.

As mentioned in one of the [comments above](https://github.com/matomo-org/tag-manager/issues/615#issuecomment-1495249271), the initial implementation won't prompt whether to copy referenced triggers and variables. If copying to the same site, it will use the same references. When copying to a different site or container, it will automatically copy referenced triggers and variables along with the tag being copied.

**Comment by snake14** [comment]
@SW-Vincent @Neugart I just merged the code implementing the ability to copy tags. Please note that it should be in the next Matomo 5.2.0-x release. Also, in order for the copy action to be visible, you must be logged in as a superuser or have write and custom template capabilities assigned to your user. The copy option tries to reuse existing triggers and variables when copying references from the tag. If no matches exist, any dependencies will be copied along with the tag.
